tempesta_fw/http.c

```diff
@@ -813,48 +1047,46 @@ tfw_http_req_cache_cb(TfwHttpReq *req, TfwHttpResp *resp)
+
 	/*
 	 * Sticky cookie module used for HTTP session identification may send
-	 * a response to the client when sticky cookie presence is enforced and
-	 * the cookie is missing from the request.
+	 * a response to the client when sticky cookie presence is enforced
+	 * and the cookie is missing from the request.
 	 *
-	 * HTTP session can be required for the request schduling, so obtain it
-	 * first. However, req->sess still can be NULL if sticky cookies aren't
-	 * enabled.
+	 * HTTP session may be required for request scheduling, so obtain it
+	 * first. However, req->sess still may be NULL if sticky cookies are
+	 * not enabled.
 	 */
 	r = tfw_http_sess_obtain(req);
 	if (r < 0)
 		goto send_500;
-	if (r > 0) {
-		/* Response sent, nothing to do. */
-		tfw_http_conn_msg_free((TfwHttpMsg *)req);
+	if (r > 0)	/* Response sent, nothing to do. */
 		return;
-	}
 
 	if (resp) {
 		tfw_http_req_cache_service(req, resp);
 		return;
 	}
 
 	/*
-	 * Dispatch request to an appropriate server. Schedulers
-	 * should make a decision based on an unmodified request,
-	 * so this must be done before any request mangling.
+	 * Dispatch request to an appropriate server. Schedulers should
+	 * make a decision based on an unmodified request, so this must
+	 * be done before any request mangling.
 	 *
-	 * The code below is typically called on remote NUMA node.
-	 * That's not good, but we must run TDB lookup on the node
-	 * before this is executed, to avoid unnecessary work in
-	 * SoftIRQ and to speed up the cache operation.
-	 * At the same time, cache hits are expected to prevail
-	 * over cache misses, so this is not a frequent path.
+	 * The code below is usually called on a remote NUMA node. That's
+	 * not good, but TDB lookup must be run on the node before it is
+	 * executed, to avoid unnecessary work in SoftIRQ and to speed up
+	 * the cache operation. At the same time, cache hits are expected
+	 * to prevail over cache misses, so this is not a frequent path.
 	 *
-	 * TODO #593: check whether req->sess->srv_conn is alive or
-	 * get a new connection for req->sess->srv_conn->peer from appropriate
-	 * scheduler otherwise. This eliminates long generic scheduling work
-	 * flow. When a first request in the session is scheduled by the generic
-	 * logic, TfwSession->srv_conn must be initialized by poniter to
-	 * appropriate TfwConnection, so all following session hits will be
-	 * scheduled much faster.
+	 * TODO #593: check whether req->sess->srv_conn is alive. If not,
+	 * then get a new connection for req->sess->srv_conn->peer from
+	 * an appropriate scheduler. That eliminates the long generic
+	 * scheduling work flow. When the first request in a session is
+	 * scheduled by the generic logic, TfwSession->srv_conn must be
+	 * initialized to point at the appropriate TfwConnection, so that
+	 * all subsequent session hits are scheduled much faster.
 	 */
 	srv_conn = tfw_sched_get_srv_conn((TfwMsg *)req);
```

Comment:
Please consider this [article about QoS](https://www.igvita.com/2008/05/13/load-balancing-qos-with-haproxy/). It has sense to add new configuration option for server group which specifies the queue size. If all the server connections are busy (`tfw_server_queue_full()` returns true) we should put a request to the new queue of configured size. If `srv_conn->qsize` decrements, then we can schedule next request from head of the new queue. If the new queue is full, then just do as currently - send error response to a client (however, it should be 50x response instead of 40x response.